### PR TITLE
[hail] Fix memory leak in BM and make cache size configurable

### PR DESCRIPTION
**Pull request by danking**
CHANGELOG: Remove memory leak in `BlockMatrix.to_matrix_table_row_major` and `BlockMatrix.to_table_row_major`. Also, make the cache size used in both methods configurable.

The `ref` variable was holding entire blocks in memory for no reason. It was a
vestiage of debugging. Moreover, the configurable cache permits users to fine tune
memory usage.

### CODE REVIEW

**Review by johnc1231 — CHANGES_REQUESTED** [review]
Docs fixes, otherwise good.

File: hail/python/hail/linalg/blockmatrix.py [hunk 1]
```diff
@@ -1699,6 +1699,14 @@ def to_table_row_major(self, n_partitions=None):
         ----------
         n_partitions : int or None
             Number of partitions of the table.
+        maximum_cache_memory_in_bytes : int or None
+            The amount of memory to reserve, per partition, to cache rows of the
+            matrix in memory. This value must be at least large enough to hold
+            one row of th matrix in memory. If this value is exactly the size of
+            one row, then a partition makes a network request for every row of
+            every block. Larger values reduce the number of network requests. If
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hail/linalg/blockmatrix.py:1707]: I fixed the verbiage. Latter should have said one partition.

**Comment by johnc1231** [comment]
KING tests are failing

**Review by johnc1231 — CHANGES_REQUESTED** [review]
Tests are failing

**Review by johnc1231 — APPROVED** [review]

**Pull request merged**